From f8d337f3d9fbbb8ca12461bd5a7fcec011cde94f Mon Sep 17 00:00:00 2001
From: Harsh J <harsh@cloudera.com>
Date: Thu, 27 Aug 2015 16:22:48 +0530
Subject: [PATCH 1897/2848] HDFS-2390. dfsadmin -setBalancerBandwidth does not
 validate -ve value. Contributed by Gautam
 Gopalakrishnan.

(cherry picked from commit 0bf285413f8fcaadbb2d5817fe8090f5fb0d37d9)

(cherry picked from commit 574f3a98ce8b8c0a55dfabc668320dc57a26a087)

Change-Id: I3d817982573ebe24f7fba4568bfc70dc5df66275
---
 .../org/apache/hadoop/hdfs/tools/DFSAdmin.java     |    5 +++++
 .../hadoop/hdfs/tools/TestDFSAdminWithHA.java      |    7 +++++++
 2 files changed, 12 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
index 2cb275c..cc3d7c1 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
@@ -799,6 +799,11 @@ public int setBalancerBandwidth(String[] argv, int idx) throws IOException {
       return exitCode;
     }
 
+    if (bandwidth < 0) {
+      System.err.println("Bandwidth should be a non-negative integer");
+      return exitCode;
+    }
+
     FileSystem fs = getFS();
     if (!(fs instanceof DistributedFileSystem)) {
       System.err.println("FileSystem is " + fs.getUri());
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestDFSAdminWithHA.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestDFSAdminWithHA.java
index 2b058d2..74f5e7a 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestDFSAdminWithHA.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestDFSAdminWithHA.java
@@ -194,6 +194,13 @@ public void testSetBalancerBandwidth() throws Exception {
   }
 
   @Test (timeout = 30000)
+  public void testSetNegativeBalancerBandwidth() throws Exception {
+    setUpHaCluster(false);
+    int exitCode = admin.run(new String[] {"-setBalancerBandwidth", "-10"});
+    assertEquals("Negative bandwidth value must fail the command", -1, exitCode);
+  }
+
+  @Test (timeout = 30000)
   public void testMetaSave() throws Exception {
     setUpHaCluster(false);
     int exitCode = admin.run(new String[] {"-metasave", "dfs.meta"});
-- 
1.7.9.5

